Optimization of the Cython Lomb-Scargle periodogram implementations

kkrutkowski opened the pull request:
PR #17244 reopened from different branch

<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Notes to maintainer

Please use the **Squash and Merge** button when this is ready for merge. Thanks!

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

The modifications can be divided into two different minor modifications;

1. Both of the Cython methods have been merged and rewritten to avoid redundant recomputation of trigonometric functions.
2. The 'assume_regular_frequency' parameter is passed to the Cython method. If true activates further optimizations limited to regular frequency grids only

Generally the Cython implementation after modifications seems to be ~ 50% faster for nonregular grids and ~ 4x  for regular ones. I've not been able to measure any significant discrepancies between results given by the old one and after that (with < 10^-12 discrepancies for MACHO dataset, just like between old Cython and Slow methods), so it shouldn't affect anything else.

kkrutkowski committed:
Pass 'assume regular frequency' to the cython implementation

Pass 'assume regular frequency' to the cython implementation as a keyword argument

kkrutkowski committed:
Major changes in cython code

Both Cython methods are now merged into one.
Redundant trigonometric computations are now skipped.
For regular grids frequency is increased recursively to speed-up the computation.

kkrutkowski committed:
Add changelog

dhomeier commented:
> Generally the Scipy implementation after modifications seems to be ~ 50% faster for nonregular grids and ~ 4x for regular ones.

Is this a typo, or how do these changes affect `lombscargle_scipy`?
It is worth noting however that the scipy function now supports a wider set of functionality enabled in #17170, so those should probably be re-reviewed in terms of performance again. But that will only be possible in the 7.1 branch after a feature freeze of scipy 1.15.

dhomeier commented:
The last commit pulling in a cosmology PR also seems unrelated – a rebasing artefact?

kkrutkowski commented:
> > Generally the Scipy implementation after modifications seems to be ~ 50% faster for nonregular grids and ~ 4x for regular ones.
> 
> Is this a typo, or how do these changes affect `lombscargle_scipy`?
> It is worth noting however that the scipy function now supports a wider set of functionality enabled in #17170, so those should probably be re-reviewed in terms of performance again. But that will only be possible in the 7.1 branch after a feature freeze of scipy 1.15.

It doesn't affect SciPy, only Cython (which is faster by the mentioned numbers). Typo fixed, thanks for noticing.

kkrutkowski commented:
> The last commit pulling in a cosmology PR also seems unrelated – a rebasing artefact?

It is, also it doesn't add any code changes to this PR. I've pulled it to run all 56 checks, as originally it run on limited subset (~ 22 i think), so after adding new flags I needed to synchronize it to relaunch the ones which were supposed to pass, but were skipped due to change of flags after reopening.

dhomeier commented on astropy/timeseries/periodograms/lombscargle/implementations/main.py:
```diff
@@ -218,6 +222,10 @@ def lombscargle(
 
+    # if the grid is regular enable additional optimizations in the cython method
+    if method == "cython":
+        kwds.update(assume_regular_frequency=assume_regular_frequency)
```
Wondering if this could additionally be checked using `_is_regular` even if not set explicitly, similar to how it's done in the 'fast' methods.

kkrutkowski replied:
It definitely could have been done, but I have no idea if it's worth it. Implicit check introducesna noticeable (but not too significant) overhead for nonregular grids, while that one generally tries to keep a good balance between uniform and nonuniform grids.

But again nonuniformly spaced ones have no practical purpose, so maybe the implicit check should've been added regardless. What do you think about it?

Btw branching in the Cython code is adding a noticeable overhead over the near-optimal implementation regardless, so it might actually be a good idea.

dhomeier commented:
The scipy numbers themselves have changed, or are about to change though. But I think actual performance figures have changed very little, so that might make the cython method clearer faster with this change.

kkrutkowski committed:
Add an implicit regularity check to the Cython implementation

kkrutkowski committed:
Make boolean parameters constant within Cython block

kkrutkowski commented:
When already adding implicit regularity check I've also added constant boolean parameters and typed indices, which generally result in noticeable speed-up for regular grids (~ 2-6%) but slightly slow down the computation for nonregular ones (~ 0.5% overhead on average), at least for Ryzen 4600h and Linux (I've changed it because the implicit detection results in similar overhead, so it should be acceptable at this point).

Btw everything benchmarked on my machine as the Repo's benchmark does not include the Cython implementation.

kkrutkowski committed:
Remove 'if' statement confusing the branch loader

neutrinoceros commented on docs/changes/timeseries/17251.perf.rst:
```diff
@@ -0,0 +1,2 @@
+Reduction of redundant computation of the the ``cython`` implementation of the Lomb-Scargle periodogram.
```
Typo. Also, we generally ask for these changelog fragment to be 80 cols wide at most
```suggestion
Reduction of redundant computation of the ``cython`` implementation of the Lomb-Scargle periodogram.
```

neutrinoceros reviewed (commented):
If I understand correctly `method="cython"` is already supported on `main` but is lacking documentation in `def lombscargle`. However what's still missing (seems to me) are tests to exercise it, and I think this is a very nice opportunity to add some.
I'm also not an expert in this algorithm(s) so having tests could help me review more confidently.

kkrutkowski commented on astropy/timeseries/periodograms/lombscargle/implementations/cython_impl.pyx:
```diff
@@ -74,15 +76,13 @@ def lombscargle_cython(t, y, dy, frequency, normalization='standard',
     # pre-center the data: not technically required if fit_mean=True,
     # but it simplifies the math.
+    w = dy ** -2
+    w /= np.sum(w)
```
Sorry, but I'm not an expert in the tests, however, the main issue with the Cython method is that it's not compiled by default during the test execution, so it's not available when the tests are running.

I would be glad if anyone could suggest a potential solution to the issue without the main redesign of the testing unit, but for now, I'll probably leave it in its current form.

One idea I have (which isn't elegant but could work) would be to import the Cython method using the [pyximport](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#compiling-with-pyximport) during the testing phase, which would work, but again would add a lot of overhead to the tests unit.

kkrutkowski replied:
> doesn't this affect the result on line 83 ? (I realize this could be balanced out by another change downstream in `_lomb_scargle`, but I want to make sure it is)

And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in `_lomb_scargle`, essentially leading to *almost* the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.

It could've been fixed with Kahan's summation or other similar algorithm, but I guess it's not needed, as it still reliably falls within the accuracy bound defined in the tests unit.

neutrinoceros replied:
> And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in _lomb_scargle, essentially leading to almost the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.

Is it *needed* here or is it just nice-to-have™️ ? It sounds to me like this could be its own separate PR (and probably should)

about testing: it's a shame that the extension isn't available in CI builds, but it shouldn't stop us from writing a test now. Tests can be conditionally skipped if their requirements are missing using `@pytest.mark.skipif`. Don't hesitate to reach out if you need more guidance.

kkrutkowski replied:
> > And about that change - it does affect the result on line 83, but it's supposed to. Essentially it pre-normalizes weights, which allows for the removal of weight recalculation during every iteration of the main loop (what the current code does). So it offsets the lack of secondary normalization in _lomb_scargle, essentially leading to almost the same results, however, it marginally increases the numerical errors, as the weights in the previous iterations were recalculated.
> 
> Is it _needed_ here or is it just nice-to-have™️ ? It sounds to me like this could be its own separate PR (and probably should)
> 
> about testing: it's a shame that the extension isn't available in CI builds, but it shouldn't stop us from writing a test now. Tests can be conditionally skipped if their requirements are missing using `@pytest.mark.skipif`. Don't hesitate to reach out if you need more guidance.

Thanks for your reply. It's just a nice-to-have™️ but an industry standard implemented in all Astropy's Lomb-Scargle implementations except the Cython. For example in the 'slow' implementation lines 53 and 54 look like that: 
`w = dy**-2.0
w /= w.sum()`
    
While rewriting the code I've tried to keep the style relatively consistent with the rest of the implementations, as Cython clearly stood out, but I could also split it if it's required. Now it looks essentially almost like the 'fast' implementation, however with all of the approximations removed to keep the precision required from all of the non-fast methods.

As for testing the biggest issue is that the tests for Cython is already implemented. It's essentially tested in one of the iterations of the loop iterating over all available methods;
`ALL_METHODS_NO_AUTO = [method for method in ALL_METHODS if method != "auto"]`

The issue with that part of the code is that the Cython is only appended after compilation (and currently it's not compiled for the tests).

The approach I wanted to implement was to 'forcefully' append the Cython to the methods even if it's not compiled by adding a code block like that;
`    import pyximport
    pyximport.install(reload_support=True, language_level=3, setup_args={"include_dirs": [np.get_include()]})
    from ..implementations import cython_impl as cython_tmp
    # Append the Cython method to available methods if import succeeds
    ALL_METHODS = LombScargle.available_methods + ["cython_tmp"]`
    
 However, even that method seemingly failed, as even pyximport was not available during the tests. So I think leaving the tests how they are (by which I mean appending cython to the list only if it's compiled) looks like the best option, at least for now. If you have any idea how could it be improved I'll gladly accept any suggestions btw.

neutrinoceros replied:
> currently it's not compiled for the tests

actually I double checked that and I'm not sure what you mean here. It looks to me that `"cython"` is unconditionally included in `ALL_METHODS_NO_AUTO` (only `"auto"` is excluded, as well as `"scipy"` *if* it's missing). We *do* compile all extensions in CI, and that's also the normal thing to do if you're running tests locally. Are you not able to run these tests locally for some reason ?

kkrutkowski replied:
Sorry for the misunderstanding then. The issue I have both locally and in the repo is the line 193 of 'all optional dependecies' tests, which returns `Cython: not available` and doesn't compile the Cython implementation from the source. If I try to state it explicitly, at least locally the test fails, as Cython is not found.

But that may be either me misunderstanding the tests module (and the `Cython: not available` in full optional dependencies test run) or some weird kind of bug (and I have no clue what may be causing it in that case).

I mean I may try temporarily changing the Cython result to something wrong (for example to set the result to be equal to 1 for every frequency) and check if it'll pass tests in repo.

At least locally it does with no issues, even if Cython returns clearly wrong result. + that line 193 seems to explain that behavior, at least if it's not a local bug of my repo copy.

kkrutkowski replied:
Thanks, then I have no idea why locally the test always passes, even with wrong results, I could've messed up my local environment in some reason then (?).

Shall I try to run tests on the Cython version intentionally returning wrong results to see, if it'll fail (like it should've) or pass like my local copy?

neutrinoceros replied:
when editing cython code, you need to re-compile astropy for your change to take effect, otherwise tests will just import the latest compiled version of the extension module. Hope this helps !

kkrutkowski committed:
Move cimport into top level of the module

kkrutkowski commented on astropy/timeseries/periodograms/lombscargle/implementations/cython_impl.pyx:
```diff
@@ -98,155 +98,92 @@ def lombscargle_cython(t, y, dy, frequency, normalization='standard',
     return PLS.ravel()
 
-
+from libc.stdlib cimport malloc, free
```
Already fixed, I'll fix the rest soon. Thx

kkrutkowski committed:
Update docs/changes/timeseries/17251.perf.rst

Co-authored-by: Clément Robert <cr52@protonmail.com>

kkrutkowski commented:
Sorry for spam due to accidentally requesting 8 reviews while trying to squash-rebase the commits.

This happened automatically as a result of clicking "Sync forc" and later removing the commits by rebasing (and I have no idea how to cancel that).

Also I have no idea why building the docs fails now, if while rebasing I didn't change anything since the original pre-rebase version.

> Hi humans 👋 - this pull request hasn't had any new commits for approximately 4 months. **I plan to close this in 30 days if the pull request doesn't have any new commits by then.**
> 
> In lieu of a stalled pull request, please consider closing this and open an issue instead if a reminder is needed to revisit in the future. Maintainers may also choose to add **keep-open** label to keep this PR open but it is discouraged unless absolutely necessary.
> 
> If this PR still needs to be reviewed, as an author, you can rebase it to reset the clock.
> 
> _If you believe I commented on this pull request incorrectly, please report this [here](https://github.com/pllim/action-astropy-stalebot/issues)._

astrofrog commented:
@kkrutkowski - sorry for dropping the ball on this, is this still needed or superseded by your other recently-merged PR?

kkrutkowski commented:
It depends what you mean by "needed".

It's just a minor reformulation of Cython implementation, which is practically always slower than fast.

Still we have prefer_fast argument or something like that (I'll have to check that one later) - and Cython is the fastest out of the "slow" implementations.

So I have no idea. I've just dropped that one because of Slack thread I have had a long time ago.

pllim commented:
There is code conflict now, so would need manual resolution if you are interested to pursue this. Thanks, all!

astrofrog committed:
Use Py_ssize_t indices, declare the temporaries, check the malloc results, guard the single-frequency case, and tidy formatting in the cython Lomb-Scargle kernel

astrofrog committed:
Add tests covering the cython Lomb-Scargle method on regular and irregular frequency grids

astrofrog committed:
Apply ruff formatting to the cython Lomb-Scargle test

astrofrog commented:
I finally got a chance to take a look at this - I went ahead and rebased it and made some small tweaks based on the review and added some more tests. I'll do a final review of this once the CI is passing.

astrofrog reviewed (approved):
This seems good to me now, I'll leave it open a few days in case anyone else has comments, otherwise let's merge.

pllim commented:
Is this related?

| Change   | Before [c5ab2223]    | After [dd8cb4f8]    |   Ratio | Benchmark (Parameter)                                      |
|----------|----------------------|---------------------|---------|------------------------------------------------------------|
| +        | 218±2μs              | 293±2μs             |    1.35 | units.TimeQuantityOpLargeArraySameUnit.time_quantity_equal |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.

kkrutkowski commented:
> Is this related?
> 
> | Change   | Before [c5ab2223]    | After [dd8cb4f8]    |   Ratio | Benchmark (Parameter)                                      |
> |----------|----------------------|---------------------|---------|------------------------------------------------------------|
> | +        | 218±2μs              | 293±2μs             |    1.35 | units.TimeQuantityOpLargeArraySameUnit.time_quantity_equal |
> 
> SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
> PERFORMANCE DECREASED.

Looks like noise to me. This PR doesn't seem to affect the code executed during that benchmarks in any ways (?), but a confirmation would be nice.

Either way the benchmark did fail a few times on this PR already in totally random places so I doubt this time it is any different.

astrofrog commented:
Yes that definitely looks unrelated
